frontend/app/src/services/history/types.ts

```diff
@@ -1,39 +1,87 @@
 import { NumericString, BigNumber } from '@rotki/common';
-import { z } from 'zod';
-import { SupportedTradeLocation } from '@/services/balances/types';
-import { TradeEntry } from '@/store/history/types';
-import { Nullable } from '@/types';
-import { SupportedExchange } from '@/types/exchanges';
+import { z, ZodTypeAny } from 'zod';
+import { SUPPORTED_TRADE_LOCATIONS } from '@/data/defaults';
+import { SUPPORTED_EXCHANGES } from '@/types/exchanges';
 
-export type TradeType = 'buy' | 'sell';
+// Common wrapper function
+const EntryMeta = z.object({
+  ignoredInAccounting: z.boolean()
+});
 
-export interface Trade {
-  readonly tradeId: string;
-  readonly timestamp: number;
-  readonly location: TradeLocation;
-  readonly baseAsset: string;
-  readonly quoteAsset: string;
-  readonly tradeType: TradeType;
-  readonly amount: BigNumber;
-  readonly rate: BigNumber;
-  readonly fee?: Nullable<BigNumber>;
-  readonly feeCurrency?: Nullable<string>;
-  readonly link?: Nullable<string>;
-  readonly notes?: Nullable<string>;
-}
+export type EntryMeta = z.infer<typeof EntryMeta>;
 
-export type NewTrade = Omit<Trade, 'tradeId' | 'ignoredInAccounting'>;
+export type EntryWithMeta<T> = {
+  readonly entry: T;
+} & EntryMeta;
 
-export interface TradeUpdate {
-  readonly trade: TradeEntry;
-  readonly oldTradeId: string;
+function getEntryWithMeta(obj: ZodTypeAny) {
+  return z
+    .object({
+      entry: obj
+    })
+    .merge(EntryMeta);
 }
 
-export type TradeLocation =
-  | SupportedExchange
-  | SupportedTradeLocation
-  | 'gitcoin';
+function getCollectionResponseType(obj: ZodTypeAny) {
+  return z.object({
+    entries: z.array(obj),
+    entriesFound: z.number(),
+    entriesLimit: z.number(),
+    entriesTotal: z.number()
+  });
+}
+
+// Trades
+export const TradeType = z.enum(['buy', 'sell']);
+export type TradeType = z.infer<typeof TradeType>;
+
+// @ts-ignore
+export const TradeLocation = z.enum([
+  ...SUPPORTED_EXCHANGES,
+  ...SUPPORTED_TRADE_LOCATIONS,
+  'gitcoin'
+]);
+
+export type TradeLocation = z.infer<typeof TradeLocation>;
+
+export const Trade = z.object({
+  tradeId: z.string(),
+  timestamp: z.number(),
+  location: TradeLocation,
+  baseAsset: z.string(),
+  quoteAsset: z.string(),
+  tradeType: TradeType,
+  amount: NumericString,
+  rate: NumericString,
+  fee: z.nullable(NumericString).optional(),
```

Comment:
I think here you can also chain the `nullable` so `NumericString.nullable().optional()`